File: programmers_Hash/hash_Whodontfinish.py
def solution(participant, completion):
    # 1. 두 list를 sort 
    participant.sort(); completion.sort();
    # 2. completeion list의 len만큼 participant를 찾아서 없는 사람을 찾는다 
    # 이 경우엔 AABC ABC 일 경우를 잡아낼 수 있음
    for i in range(len(completion)): 
        if(participant[i] != completion[i]): 
            return (participant[i])
    # 3. 전부 다 돌아도 없을 경우에는 마지막 주자가 완주하지 못한 선수이다.
    # 2번으로도 잡히지 않은 경우 : ABCD ABC (ABC까지 모두 같으므로 for문을 통과한다)
    return participant[len(participant)-1]

File: programmers_Hash/hash_Whodontfinish.cpp
/*
해시-완주하지 못한 선수
*/

#include <string>
#include <vector>
#include <unordered_map>

using namespace std;

string solution(vector<string> participant, vector<string> completion) {
    string answer = "";
    unordered_map<string, int> strMap;
    //map은 두 개의 인자를 가지는데, 첫 인자는 Key, 두 번째 인자는 Value 의 역할
    for (auto elem : completion) //auto 키워드를 사용하여 컴파일러가 자료형을 추론하도록
    {
        if (strMap.end() == strMap.find(elem))
            //end()는 빈 iterator의 주소 반환 find(a)의 경우 a를 찾지 못했으면 빈 iterator 주소 반환
            //그러므로 end()와 find(a)의 값이 같을 경우 = a의 값을 찾지 못했을 경우
            strMap.insert(make_pair(elem, 1));
        //insert(make_pair(key,value)) : pair 형태로 추가 (반드시)!!
        else
            strMap[elem]++;
    }

    for (auto elem : participant)
    {
        if (strMap.end() == strMap.find(elem)) { return elem; }
        else
        {
            strMap[elem]--;
            if (strMap[elem] < 0) { return elem; }
        }
    }
    return answer;
}